File: storage/data_transpiled/at_coder/abc104/D/4908176.java
public static void input ( ) {
  String s = input ( ) ;
  long [ ] a = {
    1 , 0 , 0 , 0 }
    ;
    for ( int i = 0 ;
    i < s . length ( ) ;
    i ++ ) {
      if ( s . charAt ( i ) == 'A' ) a [ 1 ] += a [ 0 ] ;
      else if ( s . charAt ( i ) == 'B' ) a [ 2 ] += a [ 1 ] ;
      else if ( s . charAt ( i ) == 'C' ) a [ 3 ] += a [ 2 ] ;
      else a = new long [ 3 ] ;
      for ( int j = 0 ;
      j < 3 ;
      j ++ ) a [ j ] = a [ j ] % 1000000007 ;
    }
    System . out . println ( a [ 3 ] ) ;
  }
  

File: storage/data_original/at_coder/abc104/D/4908176.py
s=input()
a=[1,0,0,0]
for i in s:
    if i=="A": a[1]+=a[0]
    elif i=="B": a[2]+=a[1]
    elif i=="C": a[3]+=a[2]
    else: a=[ a[0]*3 ] + [ a[j+1]*3+a[j] for j in range(3) ]

    a=[a[i]%1000000007 for i in range(4)]
print(a[3])